### PR TITLE
Madison/mlm

### PR DESCRIPTION
**Pull request by madisonmay**

### CODE REVIEW

File: finetune/model.py [hunk 1]
```diff
@@ -73,6 +73,33 @@ def language_model_op(X, M, params, featurizer_state):
...
+            hidden=featurizer_state["sequence_features"],
+        )
+
+        lm_logits = language_model_state["logits"]
+        lm_logit_mask = np.zeros(
+            [1, lm_logits.get_shape().as_list()[-1]], dtype=np.float32
+        )
+        lm_logit_mask[:, encoder.vocab_size :] = -np.inf
+
+        if "use_extra_toks" in params and not params.use_extra_toks:
```
Comment (benleetownsend) [inline finetune/model.py]: Is this stuff really necessary? Can we actually decode from the model? If not we should probably just remove this

File: finetune/saver.py [hunk 2]
```diff
@@ -258,7 +258,9 @@ def init_fn(scaffold, session, model_portion=None, refresh_base_model=False):
                     saved_var = variables_sv[name]
                 elif name in self.fallback.keys():
                     saved_var = self.fallback[name]
-
+                else:
+                    if 'OptimizeLoss' not in name:
```
Comment (benleetownsend) [inline finetune/saver.py]: I think you accidentally left this in. But Maybe we can do some variant of this always? Or behind some flag?

File: finetune/model.py [hunk 3]
```diff
@@ -174,10 +197,21 @@ def _model_fn(features, labels, mode, params):
...
+                elif lm_type.lower() == 'mlm':
+                    lm_predict_op, language_model_state = masked_language_model_op(
+                        X=X, 
+                        M=M, 
+                        mlm_weights=features['mlm_weights'],
+                        mlm_ids=features['mlm_ids'],
+                        mlm_positions=features['mlm_positions'],
+                        params=params, 
+                        featurizer_state=featurizer_state
+                    )
```
Comment (benleetownsend) [inline finetune/model.py]: else: raise an error?

File: finetune/nn/target_blocks.py [hunk 4]
```diff
@@ -30,6 +31,62 @@ def perceptron(x, ny, config, w_init=None, b_init=None):
...
+    with tf.variable_scope('model/masked-language-model'):
+        gathered_hidden = gather_indexes(hidden, mlm_positions)
+        final_proj = tf.layers.dense(
+            gathered_hidden,
+            units=config.n_embed,
+            activation=act_fns[config.act_fn],
+            kernel_initializer=tf.random_normal_initializer(stddev=config.weight_stddev),
+            name='dense'
+        )
+        normed_proj = norm(final_proj, 'LayerNorm')
```
Comment (benleetownsend) [inline finetune/nn/target_blocks.py]: If we have the choice here I think its best to use tensorflow's built in layer norm like bert does.

File: finetune/scripts/rename_bert_mlm_weights.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import joblib
```
Comment (benleetownsend) [inline finetune/scripts/rename_bert_mlm_weights.py]: Are we going to start keeping these kinds of scripts in finetune? If so I will add in the ones to port BERT models etc

File: finetune/model.py (hunk 1 above)
Comment (madisonmay) [reply finetune/model.py]: Yeah we don't need that. I still need to refactor this chunk.  Got distracted dealing with missing base model weights and vocab entries yesterday :/

File: finetune/saver.py (hunk 2 above)
Comment (madisonmay) [reply finetune/saver.py]: Yeah honestly not a bad idea.  Could make a shortlist of weights that should be randomly initialized and then raise a warning if anything else is randomly init-ed.

File: finetune/model.py (hunk 3 above)
Comment (madisonmay) [reply finetune/model.py]: :+1:

File: finetune/nn/target_blocks.py (hunk 4 above)
Comment (madisonmay) [reply finetune/nn/target_blocks.py]: Yeah that would be fine.  Should be equivalent though right?

File: finetune/scripts/rename_bert_mlm_weights.py (hunk 5 above)
Comment (madisonmay) [reply finetune/scripts/rename_bert_mlm_weights.py]: I added them temporarily for purposes of review so you can interpret other code changes w/ these changes in mind.  I think generically we should have done this but things are scattered all over the place right now.

**Comment by madisonmay** [comment]
- Uploading model files now
- Confirmed that old bert / roberta model files load and produce the same responses when moving from master to this branch

Still need to:
- rebase
- run tests

**Comment by madisonmay** [comment]
At Lily's suggestion going to take some quick notes on this PR for future reference:
Had to update the BERT base model files because although we have saved MLM target model weights in our current save files -- they're scoped differently than the rest of finetune.  So I changed the keys and re-uploaded -v2 model files.
Same goes for RoBERTa but we also have some additional complexity because our vocab files are *also* missing some additional entries.  To handle this I create a RobertaEncoderV2 class that points to new model files, and select the proper encoder to use based on the value of base_model_path. 
Finally, it would also be sensible to add support for DistilBERT models to be trained using MLM objective, but we consider that outside the scope of this PR and have them simply error out for now.

File: finetune/target_models/masked_language_model.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+                    "tokens": TS([self.config.max_length, 2]),
+                    "mask": TS([self.config.max_length]),
+                    "mlm_weights": TS([self.config.max_masked_tokens * self.config.batch_size]),
+                    "mlm_ids": TS([self.config.max_masked_tokens * self.config.batch_size]),
+                    "mlm_positions": TS([self.config.max_masked_tokens * self.config.batch_size]),
+                },
+            ),
+        )
+
+    def text_to_tokens_mask(self, X, Y=None):
```
Comment (lhz1029) [inline finetune/target_models/masked_language_model.py]: Here's one point of interaction between context and mlm. The other one currently is context and chunk_long_sequences (I hope to fix that soon, but ideally not block merging the current stuff for this?)

I added a quick line in the docs in my pr.

**Pull request merged**